### PR TITLE
Feat/capture time delay 20

### PR DESCRIPTION
**Pull request by laurawheatley**
Implemented the backend functionality for the [Capture Time Delay feature](https://github.com/aa-software2112/SOEN390_SimpleCamera/issues/20) [#20].

### CODE REVIEW

**Comment by KrishnaPatel1** [comment]
@laurawheatley @Lussier115 @samgo1 @lkadian  Something went wrong in the testing stage of the CI build. I think it timed out or something like that. Needs to be investigated further

File: app/src/main/kotlin/com/simplemobiletools/camera/activities/MainActivity.kt [hunk 1]
```diff
@@ -222,6 +228,16 @@ class MainActivity : SimpleActivity(), PhotoProcessor.MediaSavedListener {
         settings.setOnClickListener { launchSettings() }
         toggle_photo_video.setOnClickListener { handleTogglePhotoVideo() }
         change_resolution.setOnClickListener { mPreview?.showChangeResolutionDialog() }
+        countdown_toggle.setOnClickListener { toggleCountdownTimer() }
+
+        // TODO
+        // - Right now this is somewhat hard-coded
+        // - May want to look into using Android Spinner instead (similar to dropdowns)
+        // - The 3 lines below should be 1 that takes the value from front end to set the delay
+        // - https://developer.android.com/guide/topics/ui/controls/spinner
```
Comment (KrishnaPatel1) [inline app/src/main/kotlin/com/simplemobiletools/camera/activities/MainActivity.kt:237]: Great comment here!

File: app/src/main/kotlin/com/simplemobiletools/camera/activities/MainActivity.kt [hunk 2]
```diff
@@ -230,6 +246,43 @@ class MainActivity : SimpleActivity(), PhotoProcessor.MediaSavedListener {
...
+    }
+
+    private fun unsetCountdownMode() {
+        mCountdownTime = 0
+        mIsInCountdownMode = false
+        toggleCountdownModeIcon(mCountdownTime)
+    }
+
+    private fun toggleCountdownTimer() {
+
```
Comment (KrishnaPatel1) [inline app/src/main/kotlin/com/simplemobiletools/camera/activities/MainActivity.kt]: I'd remove this empty line-break at line 264

**Comment by lkadian** [comment]
@KrishnaPatel1 Build wasn't failing before I merged develop into this branch to resolve conflicts, so that problem isn't related to the feature. I'll look into it when I have time though.

**Comment by KrishnaPatel1** [comment]
So if I understand correctly, the dropdown for the timer settings isn't actually an Android countdown? It's a hardcoded piece of frontend?

File: app/src/main/kotlin/com/simplemobiletools/camera/activities/MainActivity.kt [hunk 3]
```diff
@@ -435,6 +527,29 @@ class MainActivity : SimpleActivity(), PhotoProcessor.MediaSavedListener {
         })
     }
 
+    // TODO: May want to put this in CameraPreview.kt; next to tryTakePicture()
+    private fun tryTakeDelayedPicture() {
```
Comment (KrishnaPatel1) [inline app/src/main/kotlin/com/simplemobiletools/camera/activities/MainActivity.kt:530]: Like you said, I think it belongs in CameraPreview.kt. I recommend doing it before merging if it isn't too much work. Or else, I'll make a task that can be completed during sprint 3.

**Comment by Lussier115** [comment]
Seems to be the last commit that made it crash. Nothing to do with the front end

**Comment by lkadian** [comment]
> So if I understand correctly, the dropdown for the timer settings isn't actually an Android countdown? It's a hardcoded piece of frontend?

Yes, however after looking at it again I'm not even sure we want to use a drop down, because the way we implement it and make it look on the UI it's not really a drop down. So I think it's fine to leave as is, works well anyways.

**Review by Lussier115 — COMMENTED** [review]
Why this change?

**Comment by KrishnaPatel1** [comment]
@lkadian I agree, it looks nicer as is. A dropdown would've made it look too much like a menu item.

**Comment by Lussier115** [comment]
@lkadian @KrishnaPatel1  We good

**Comment by KrishnaPatel1** [comment]
Unless @aa-software2112 has anything to say, go ahead with the merge. Good job people!

File: app/src/main/kotlin/com/simplemobiletools/camera/activities/MainActivity.kt (hunk 1 above)
Comment (aa-software2112) [reply app/src/main/kotlin/com/simplemobiletools/camera/activities/MainActivity.kt:237]: Are we doing this or not? If not, remove this comment @lkadian

**Comment by aa-software2112** [comment]
I would suggest the following for future implementations:

1) As it stands, for unit testing, if we want to test the methods that were added for the countdown timer - they are all within the MainActivity class. The issue with this is:
a) If we keep on this, we will see MainActivity grow to that "God Class" that simply has methods and functionality that it is not responsible for implementing. The MainActivity class will grow to a size that will not be manageable. Creating a countdown timer class that the MainActivity interacts with is much cleaner.
b) Later, when we are testing some other feature, we will likely want to mock the countdown timer (as it is a time-dependent system), but we cannot, as all the methods are part of the MainActivity class, so we would need to mock the MainActivity class, however we cannot do this (unless we use a 'spy') because the MainActivity class is actually the object from which all testing must begin (and we don't mock the object under test when testing its behavior). Think of the MainActivity class like the "sale" class in SOEN 345.
c) Another reason to make the countdown timer as an object, is that we can instantiate it in the MainActivity (as it is currently), and ontop of that, we can pass mocked versions with an alternate and parameterized constructor.

I'm OK to merge, we will have to keep the implementation as such, but do be careful in the future; objects are key for testing and making isolated fixes.

**Comment by lkadian** [comment]
I think we should keep the toggle/set methods even for the new feature in the MainActivity.kt, since that is where we set the event listeners and where the toggles for that activity are usually located. I'd transfer the tryTakeDelayedPicture() method to CameraPreview.kt  and also make a new class for the timer itself. This shouldn't take so long but will probably change the tests. I could work on it tonight.

**Comment by aa-software2112** [comment]
@lkadian of course the front end interaction with MainActivity needs to have the necessary function handlers, however much of it could be a class that holds some view object that it modifies internally

**Comment by lkadian** [comment]
The button toggles that simply interact with the UI should stay in MainActivity.kt because they are only relevant to that class/activity. I agree some of the functions can be moved to a new class (say CameraCountdownTimer.kt). Mainly the set/usnsetCountdownMode() function, and the tryTakeDelayedPicture() function which can be split and put into CameraPreview.kt and CameraCountdownTimer class. I'll make a final refactoring before merging.

**Pull request merged**